src/main/java/org/apache/sling/xss/impl/style/BatikCssCleaner.java

```diff
@@ -0,0 +1,82 @@
+/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ ~ Licensed to the Apache Software Foundation (ASF) under one
+ ~ or more contributor license agreements.  See the NOTICE file
+ ~ distributed with this work for additional information
+ ~ regarding copyright ownership.  The ASF licenses this file
+ ~ to you under the Apache License, Version 2.0 (the
+ ~ "License"); you may not use this file except in compliance
+ ~ with the License.  You may obtain a copy of the License at
+ ~
+ ~   http://www.apache.org/licenses/LICENSE-2.0
+ ~
+ ~ Unless required by applicable law or agreed to in writing,
+ ~ software distributed under the License is distributed on an
+ ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ ~ KIND, either express or implied.  See the License for the
+ ~ specific language governing permissions and limitations
+ ~ under the License.
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+package org.apache.sling.xss.impl.style;
+
+import java.io.IOException;
+import java.io.StringReader;
+
+import org.apache.batik.css.parser.Parser;
+import org.apache.sling.xss.impl.xml.AntiSamyPolicy.CssPolicy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.w3c.css.sac.CSSException;
+import org.w3c.css.sac.InputSource;
+
+public class BatikCssCleaner {
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+    private final CssPolicy cssPolicy;
+
+    private static final String CDATA_PRE = "<![CDATA[";
+    private static final String CDATA_POST = "]]>";
+
+    public BatikCssCleaner(CssPolicy cssPolicy) {
+        this.cssPolicy = cssPolicy;
+    }
+
+    /**
+     * Parses a CSS stylesheet and returns it in a safe form
+     *
+     * @param untrustedCss a complete CSS stylesheet
+     * @return the cleaned CSS stylesheet text
+     */
+    public String cleanStylesheet(String untrustedCss) {
+        try {
+            if ( untrustedCss.startsWith(CDATA_PRE) && untrustedCss.endsWith(CDATA_POST) )
+                untrustedCss = untrustedCss.substring(CDATA_PRE.length(), untrustedCss.length() - CDATA_POST.length());
+            Parser parser = new Parser();
+            ValidatingDocumentHandler handler = new ValidatingDocumentHandler(cssPolicy, false);
+            parser.setDocumentHandler(handler);
+            parser.parseStyleSheet(new InputSource(new StringReader(untrustedCss)));
+            return handler.getValidCss();
+        } catch (CSSException | IOException e) {
+            logger.debug("Unexpected error while cleaning stylesheet", e);
```

Comment:
Should we really silently return the empty string in case of exceptions? I would rather not catch here

Reply:
This method is called, by the [CSSValidator newCssAttributePolicy Method](https://github.com/apache/sling-org-apache-sling-xss/pull/28/files#diff-6abfffeec1aa36d3bbef5bb53c2a5ace07f61bab8c38b34a5bf11002f8457d4eR45).
And this is an Override of the[ AttributePolicy apply Method](https://github.com/OWASP/java-html-sanitizer/blob/main/src/main/java/org/owasp/html/AttributePolicy.java#L59), which I did not write.
Should i catch it instead in the CSSValidator class?

Reply:
I don’t think that catching those errors at all makes sense. Just don’t catch the unchecked exception and let the caller deal with it. This is IMHO a configuration error which should lead to a 500 response in servlets

Reply:
The caller `apply` and `text` method does not allow me to throw a checked exception.
(1) should i modify the caller method signature to throw a checked exception?
(2) or should i re-throw the `(CSSException | IOException e)` as runtime-Exception?
(3)or should i simply return an empty string and log the error?

Reply:
The caller `apply` and `text` method does not allow me to throw a checked exception.
(1) should i modify the caller method signature to throw a checked exception?
(2) or should i re-throw the (CSSException | IOException e) as runtime-Exception?
(3)or should i simply return an empty string and log the error?

Reply:
The parser errors occur due to malformed style sheets or declarations supplied to the parser. It does not mean that we have misconfigured it, the policy handling does not do that.

We have decided to return an empty output for invalid input since it is in line with the overall behaviour of the Java HTML Sanitiser. It does not throw exceptions on invalid input and the response for really malformed data e.g. `<stron|` is an empty output.

It does have an error reporting facility, via `HtmlStreamRenderer.error`, but that is used for malformed HTML and not exposed to processors that are invoked later.

I think this is pretty much in line with how the library is used and would not change it.

Reply:
But at least it should be logged with ERROR level (if you swallow such exceptions which lead to empty stylesheets). Otherwise debugging will be way too hard.

Reply:
This works for me, thanks.

Reply:
Resolved: 3ae09560240b2ef59b4bd354aeca6c3d73fa3393